incorporate pandera into network structure

heckstrahler opened the pull request:
continue work from #2717 

- ~~incorporate pandera schemata into network structure creation~~
- ~~make sure string dtypes are properly set after element creation~~
- ~~write tests for validation~~
- ~~write tests for result validation?~~
- ~~test validation for every create function (not all create functions have tests...)~~
- ~~automate docu csv creation~~ (see #2941)
- ~~CGMES tags~~
- ~~default values~~

- ~~decide on how to handle name column~~

- trafo creation needs tap variable handling, required seems wrong, tap variables should only exist together

TODO: CHANGELOG!!!

Questions:
~~switch z_ohm required~~
sc, 3ph dependancy, trafo, trafo3w

heckstrahler committed:
add pandera to dependancies

vogt31337 committed:
added missing vn_kv

vogt31337 committed:
fixed a broken check about standard values... (which are not documented!)

vogt31337 committed:
example on how metadata could be used

vogt31337 committed:
fix which will not add non required columns.

heckstrahler committed:
add descriptions to schemata, add b2b_vsc, bus_dc and load_dc schemata

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
add schema for dcline, measurement, source_dc, ward, xward

heckstrahler committed:
add schemata for bi_vsc, impedance, line_dc, tcsc, trafo3w

heckstrahler committed:
add res schema to elements

heckstrahler committed:
add 3ph res tables to elements

heckstrahler committed:
remove introduced inplace methods

heckstrahler committed:
fix if condition for dataframe

heckstrahler committed:
add metadata to elements

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

FynnIEE committed:
add additional schema (#2738)

* add motor.py

* add motor.py

* adding pa.Check to bus
adding min and max value

* adding import for motor

* adding asymmetric_sgen.py
adding import for asymmetric_sgen.py

* deleting import for asymmetric_sgen
deleting current_source from asymmetric_load
adding import for asymmetric_load
note: asymmetric_sgen was not added in the last commit (spelling mistake)

* adding asymmetric_sgen.py
adding import for asymmetric_sgen

* adding storage.py
adding import for storage

* fix doc for gen (cos_phi in csv)
adding gen.py
adding import for gen

* adding shunt.py
adding import for shunt

* fixed formatting in svc_par.csv (max_angle_degree)
adding svc.py
adding import for svc

* adding res_schema for svc
editing schema names
adding format for res_scv

* editing schema names back

* adding vsc.py
adding import for vsc
adding res_schema for vsc
adding import for res_schema

* adding res_schema for motor
adding import for res_motor
adding res_schema for asymmetric_load
adding import for res_asymmetric_load
adding res_schema for asymmetric_sgen
adding import for res_asymmetric_sgen

* adding res_schema for storage
adding import for res_storage
adding res_schema for gen
adding import for res_gen
adding res_schema for shunt
adding import for res_shunt

* deleting strict_False from res_schema

* adding 3ph for asymmetric_load
adding import for res_schema_3ph
adding 3ph for asymmetric_sgen
adding import for res_schema_3ph
adding 3ph for storage
adding import for res_schema_3ph

* adding metadata to storage
adding metadata to gen

heckstrahler committed:
fix less equals error

heckstrahler committed:
reformat imports

heckstrahler committed:
set correct dtype for id_characteristic_table, and set
tap_dependency_table to required

heckstrahler committed:
fix typo, fix small slow list/dict creation

heckstrahler committed:
make comments into TODOs

KS-HTK committed:
fixed failing sq_lite test

heckstrahler committed:
add schema for ssc element, bug fixes

heckstrahler committed:
revert breaking change, other bugfixes

KS-HTK committed:
fixed test for subnet

heckstrahler committed:
fix typing in build_branch

heckstrahler committed:
add metadata to csv creation

heckstrahler committed:
update get_std_type_structure_dict

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

KS-HTK committed:
fixed path loading
added pandera stubs to requirements

KS-HTK committed:
Merge branch 'develop' into structure_dict_rework

# Conflicts:
#	pyproject.toml

heckstrahler committed:
set required correctly for bus, bus_dc

FynnIEE committed:
Structure dict rework add additional schemas (#2745)

* add motor.py

* add motor.py

* adding pa.Check to bus
adding min and max value

* adding import for motor

* adding asymmetric_sgen.py
adding import for asymmetric_sgen.py

* deleting import for asymmetric_sgen
deleting current_source from asymmetric_load
adding import for asymmetric_load
note: asymmetric_sgen was not added in the last commit (spelling mistake)

* adding asymmetric_sgen.py
adding import for asymmetric_sgen

* adding storage.py
adding import for storage

* fix doc for gen (cos_phi in csv)
adding gen.py
adding import for gen

* adding shunt.py
adding import for shunt

* fixed formatting in svc_par.csv (max_angle_degree)
adding svc.py
adding import for svc

* adding res_schema for svc
editing schema names
adding format for res_scv

* editing schema names back

* adding vsc.py
adding import for vsc
adding res_schema for vsc
adding import for res_schema

* adding res_schema for motor
adding import for res_motor
adding res_schema for asymmetric_load
adding import for res_asymmetric_load
adding res_schema for asymmetric_sgen
adding import for res_asymmetric_sgen

* adding res_schema for storage
adding import for res_storage
adding res_schema for gen
adding import for res_gen
adding res_schema for shunt
adding import for res_shunt

* deleting strict_False from res_schema

* adding 3ph for asymmetric_load
adding import for res_schema_3ph
adding 3ph for asymmetric_sgen
adding import for res_schema_3ph
adding 3ph for storage
adding import for res_schema_3ph

* adding metadata to storage
adding metadata to gen

* adding optional parameter to the line

* adding optional parameter to the line_dc
adding metadata tdpf to line

* adding optional parameter to bus_dc

* adding optional parameter to gen

* adding optional parameter to impedance

* adding optional parameter to sgen

* adding metadata opf for bus and gen

* creating to do for in_ka in create.py
adding optional parameter to trafo

* Update gen.py

* Update impedance.py

* Update line.py

* Update gen.py

* Update trafo.py

* Update line_dc.py

---------

Co-authored-by: David Heck <david@heck.info>

heckstrahler committed:
set better variable names, correct imports, add missing res schemata

heckstrahler committed:
set required according to documentation

heckstrahler committed:
fix bus creation when there is no geo data, other schema fixes

heckstrahler committed:
make test_create as runable as possible -> update create, allow non
required from network_structure, update schemas

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
further changes for test create

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
fix errors for merge, fix dtype in _set_multiple

heckstrahler committed:
fixed nullable strings dtype, some typing, some sonarlint warnings

heckstrahler committed:
add bus index validation method

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
add foreign key for bus in schematas

heckstrahler committed:
check bus, bus_dc, dcline, ext_grid, gen, set name to pd.StringDtype

heckstrahler committed:
make res nullable

heckstrahler committed:
set nullable = true if required = False

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
add validation for columns which need to exist together

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
fix import

heckstrahler committed:
add hint to trafo check

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
update pandera definitions

heckstrahler committed:
update schemata after plausibility check

heckstrahler committed:
Merge branch 'develop' into structure_dict_rework

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix import get_dtypes

heckstrahler committed:
fixes for test_create

heckstrahler committed:
add tests for pandera get_dtypes and group_dependency, fix error in
group_dependency

heckstrahler committed:
fix dtype in set_multiple_entries, fix geodata init in bus_create, fix
dtype tests in test_create

heckstrahler committed:
fix group_dependency

heckstrahler committed:
fix various create problems

heckstrahler committed:
remove old tools file

heckstrahler committed:
update repo

heckstrahler committed:
add pandera bus schema test

heckstrahler committed:
move numbers to helper

heckstrahler committed:
fix pandera test bus

heckstrahler committed:
update schemata

heckstrahler committed:
improve test variables

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
add zero number so helper

heckstrahler committed:
fix description errors

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
update bus schema test

heckstrahler committed:
serveral schema fixes

heckstrahler committed:
fix max step for shunt schema

heckstrahler committed:
fixes for shunt and sgen schemata

heckstrahler committed:
fixes for gen and trafo runpp

heckstrahler committed:
revert build branch calc tap from dataframe problem..

heckstrahler committed:
set switch z_ohm to required

heckstrahler committed:
add column_condition, lower equals check creation method

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
add fixes for test_create

heckstrahler committed:
fix group validation error

heckstrahler committed:
fix create test

heckstrahler committed:
rename pandera test folder

heckstrahler committed:
fix some cgmes converter problems

heckstrahler committed:
fix more errors in converter

heckstrahler committed:
fix asmcsgen type

KS-HTK committed:
fixes in build_branch.py (#2804)

KS-HTK committed:
fixes for all network and plotting tests (#2805)

heckstrahler committed:
restore opf check in bus schema

heckstrahler committed:
Merge branch 'structure_dict_extension' of github.com:e2nIEE/pandapower into structure_dict_extension

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

KS-HTK committed:
fixed issues in test_auxiliary.py (#2808)

heckstrahler committed:
Merge branch 'structure_dict_extension' of github.com:e2nIEE/pandapower into structure_dict_extension

heckstrahler committed:
add lower than vor vkr and vk at trafo and trafo3w

heckstrahler committed:
add comment

heckstrahler committed:
update schemata

heckstrahler committed:
add 3ph to rx min max for ext grid

heckstrahler committed:
fix create lower than column check

heckstrahler committed:
fix problems with bus max min vmpu creation

heckstrahler committed:
fix NA from excel, current_source in sgen, discretetapcontrol comparison

heckstrahler committed:
fix run3ph when type is not set

heckstrahler committed:
fix add dcline gens

KS-HTK committed:
Issues/structure dict extension test fixes (#2824)

* fixed std_type errors

* fixed topology tests

* fixed test_nets_equal

heckstrahler committed:
fixes for grid equivalents

KS-HTK committed:
Issues/structure dict extension fix issues (#2825)

* fixed api.test_diagnostic

* fixed trafo create

* fixed issue with NA in _calculate_3w_tap_changers

* fixed use of .values in _calculate_3w_tap_changers

* fixed use isna only working on non scalar objects

* fixed issues in test_cost_consideration

heckstrahler committed:
fix bus dc opf nullable

KS-HTK committed:
Issues/structure dict extension fix issues (#2827)

* split test with for loop into single executions

* fixed order of bus table entries to avoid dtype loss

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix sgen create kappa default

KS-HTK committed:
Issues/structure dict extension fix issues (#2830)

* replaced query with direct access
%timeit net.gen.query("in_service").bus
1.02 ms ± 34 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit net.gen.where(net.gen.in_service).bus
894 μs ± 33.7 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit net.gen[net.gen.in_service]
145 μs ± 3.75 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

* fixed typo in comment

* fixed _get_slack_weights

* fixed some grid_equivalents tests
fixed gen schema and create methods

* fixed std_type containing nan for trafo

* fixed boundary_testnet using ints in string column 'name'

* fixed issue where empty series had incorrect index

* fixed issue when accessing name column while not present

* fixed creat_buses creating the elements before adding all values to entries
fixed the test_get_exivalent.test_basic_usecases test

* fixed use of is_object_dtype to determine if column can contain None/NaN/Null/NA elements.

* refactored function for better legibility

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
add phases for sn_mva in asym load/sgen

heckstrahler committed:
fix bus create for opf parameter

heckstrahler committed:
fix bus test

heckstrahler committed:
fix test_grid_equivalents_auxiliary (test_trafo_phase_shifter)

heckstrahler committed:
tap_side fix in _calculate_3w_tap_changers

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

lezneew committed:
added not_ints_list in helper.py (#2798)

* added not_ints_list in helper.py
added tscs test file

* fixed stuff

* [fixed] test_optional_fields_with_nulls comment

* [added] bus_dc pandera test

* tcsc: changed optional test and name
bus_dc: changed optional test and name
bus: changed helper path

* [added] test_pandera_dcline_elements.py

* refactor

* [added] test_pandera_asymetric_load_elements.py

* [added] test_pandera_asymetric_sgen_elements.py

* [fixed] bus schema check list

* refactored code. trafo schema issue

* [added] b2b_vsc pandera test

* [added] bi_vsc pandera test
[refactored] b2b_vsc. only has 1 optional. redundand tests

* [added] ext_grid pandera test

* [fixed] order in bus_create.py

* added schema tests

* fixed. bool type issue remains

* added schema tests

* added schema tests
refactored code

* fixed issues in trafo pandera test but still fails

* added tests

---------

Co-authored-by: David Heck <david.heck@iee.fraunhofer.de>

heckstrahler committed:
remove redundant test

heckstrahler committed:
fix sgen , line, lindc tests

heckstrahler committed:
fix storage tests

heckstrahler committed:
fix svc test

heckstrahler committed:
fix trafo and shunt tests

heckstrahler committed:
fix line test

heckstrahler committed:
fix tests

heckstrahler committed:
fix test_get_min_max_q_capability

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix for contingency

heckstrahler committed:
fix test_recycle_dc_trafo_shift

heckstrahler committed:
compare_tcsc_impedance

heckstrahler committed:
fix trafo test

heckstrahler committed:
fix test_ext_grid_gen_order_in_ppc

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix test_merge_with_characteristics

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
trafo and other smaller fixes

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix _add_dcline_gens

heckstrahler committed:
trafo and sc fixes

heckstrahler committed:
fix test_characteristic

heckstrahler committed:
fixes and sonarcube fixes

heckstrahler committed:
fix typing

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix test for change element name

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix network structre res vsc

heckstrahler committed:
rename bivsc to vscbipolar

heckstrahler committed:
add add column to df

lezneew committed:
Structure dict extension (#2882)

* added not_ints_list in helper.py
added tscs test file

* fixed stuff

* [fixed] test_optional_fields_with_nulls comment

* [added] bus_dc pandera test

* tcsc: changed optional test and name
bus_dc: changed optional test and name
bus: changed helper path

* [added] test_pandera_dcline_elements.py

* refactor

* [added] test_pandera_asymetric_load_elements.py

* [added] test_pandera_asymetric_sgen_elements.py

* [fixed] bus schema check list

* refactored code. trafo schema issue

* [added] b2b_vsc pandera test

* [added] bi_vsc pandera test
[refactored] b2b_vsc. only has 1 optional. redundand tests

* [added] ext_grid pandera test

* [fixed] order in bus_create.py

* added schema tests

* fixed. bool type issue remains

* added schema tests

* added schema tests
refactored code

* fixed issues in trafo pandera test but still fails

* added tests

* updated vsc_bipolar test to use create function
deleted old test file for bi_vsc

---------

Co-authored-by: David Heck <david.heck@iee.fraunhofer.de>

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
Merge branch 'structure_dict_extension' of github.com:e2nIEE/pandapower into structure_dict_extension

heckstrahler committed:
fix for dev merge introduced errors

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
set controallable to default false when optional

heckstrahler committed:
fix tests

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix create controllable errors

heckstrahler committed:
fix name in ucte converter

heckstrahler committed:
fix trafo.df for ucte converter

heckstrahler committed:
fix stationcontrol

heckstrahler committed:
fix sonar warnings

KS-HTK committed:
Structure dict extension fixes (#2891)

* fixed multiple issues in create

* fix of api tests

* fixed some tests

* fixed rei_generation summing bools

* fixed gird_equivalent tests
fixed loadDC casting of bool col to boolean

* fixed issue with incorrect dtypes set by cim converter.

---------

Co-authored-by: Hjalmar Koertge <hkoertge@node3.omnia.cluster>

heckstrahler committed:
fix test_lightsim2grid_phase_shifters

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

KS-HTK committed:
structure dict extension more fixes (#2892)

* fixed multiple issues in create

* fix of api tests

* fixed some tests

* fixed rei_generation summing bools

* fixed gird_equivalent tests
fixed loadDC casting of bool col to boolean

* fixed issue with incorrect dtypes set by cim converter.

* ignore_errors should not be True if run_powerflow could fail

* set default value for leakage_resistance_ratio_hv to 0.5

* compare default geodata with isna not isnan

* fixed typo

---------

Co-authored-by: Hjalmar Koertge <hkoertge@node3.omnia.cluster>

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

KS-HTK committed:
Issues/fixes in structure dict extension (#2900)

* formatting improvements

* fixed issue in runpp where voltage_depend_loads would not be disabled if conts params were not set

* minor improvement in code complexity

* fixed int passed as name to create_impedance

* fixed typo in table name

* readded missing line

* fixed slack_weight for gen set without default in test network
removed slack_weight on sgen: correct?

* removed todo note

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
Merge remote-tracking branch 'refs/remotes/origin/structure_dict_extension' into structure_dict_extension

heckstrahler committed:
add default values to metadata

KS-HTK committed:
Issues/fixes in structure dict extension (#2904)

* formatting improvements

* fixed issue in runpp where voltage_depend_loads would not be disabled if conts params were not set

* minor improvement in code complexity

* fixed int passed as name to create_impedance

* fixed typo in table name

* readded missing line

* fixed slack_weight for gen set without default in test network
removed slack_weight on sgen: correct?

* removed todo note

* fixed test failing because of missing columns in dcline
  changed check, nan in these columns would not fail so no column should also not fail

* fixed test_q_capability_curve failing due to missing column

* formatting and typo fixes

* added missing reactive_capability_curve column to sgen test net.

* fixed missing columns in ucte converter
minor typo fixes

* fixed shunt characteristic table diagnostic test.

* fixed trafo characteristic table diagnostic test.

* improvement in lightsim2grid error logging

* parameterized test and added skip for a case that fails from error in a dependencie

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

KS-HTK committed:
Issues/fixes in structure dict extension (#2920)

* skips for tests erroring because of issue in lightsim2grid

* added missing columns to test_qlimits_with_capability_curve test

* added column sorting to add_column_to_df

* fixed issue in test_from_mpc_mat

* updated comment in test_from_mpc.py

* fixed ucte converter not creating all required columns

* fixed ppc test

* added missing column max_vm_pu for rei test

* added missing columns min_q_mvar and max_q_mvar to sgen for BinarySearchControl

heckstrahler committed:
fix BinarySearchControl

heckstrahler committed:
introduce cim metadata, create columns for cim2pp from schemata, all networkcreation with given structure

Commit:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix _update_min_max_q_mvar for station_control

KS-HTK committed:
Fix facts structure dict extension (#2922)

* added missing column slack_weight to test_svc_tcsc_case_study test

* added temporary skip for PGM tests due to issue in their converter

* added controllable=True to gens for test_only_gen_slack_vm_setpoint

* added missing columns to gen table for test_gen_violated_p_vm_limits
minor improvement in warning for opf param check

* added missing columns to gen table for test_gen_p_vm_limits

* added missing columns to gen table for test_cost_piecewise_linear_eg

* added missing columns to gen table for test_minimize_active_power_curtailment

* added missing columns to dcline table for test_opf_task

fixed issue in validate_opf_input.py

heckstrahler committed:
fixes for cim2pp

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fix mypy

heckstrahler committed:
fix todos

KS-HTK committed:
updated docstings (#2934)

deprecated unused keyword args

heckstrahler committed:
add ucte converter tags

heckstrahler committed:
fix bool defaults

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

jwiemer112 committed:
Feature/get all elements (#2939)

* Hallo

* Hallo Leute

* adding first draft of simple_hl_plot a function to easily highlight buses and lines paired with a hover functionality

* added function to get a dataframe of all elements in the net

* add test
clean comments
renaming

* add test

* add docu

* add changelog entry

mrifraunhofer committed:
ltds_fix (#2938)

* updated ltds notebooks

* updated ltds notebooks, fixed LTDS converter

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

KS-HTK committed:
added pytest split to tutorial_test pipeline (#2935)

* added pytest split to tutorial_test pipeline

* added notebook test times

KS-HTK committed:
fixed many links (#2937)

other formatting

Co-authored-by: V3 <mike.vogt@iee.fraunhofer.de>

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
update pandera version

heckstrahler committed:
set pandera version interval

heckstrahler committed:
set pandaer 0.29.0 as max version

brynpickering committed:
Fix copyright text formatting in LICENSE file (#2914)

* Fix copyright text formatting in LICENSE file

* Refactor copyright notice in LICENSE

Updated copyright notice formatting in LICENSE file following [upstream changes](https://github.com/licensee/licensee/pull/950) to handle multi-line attribution.

---------

Co-authored-by: KS_HTK <2981026+KS-HTK@users.noreply.github.com>

mrifraunhofer committed:
added PowerElectronicsConnection to the LTDS converter, updated LTDS notebooks (PowerElectronicsConnection) (#2942)

Co-authored-by: David Heck <david.heck@iee.fraunhofer.de>

KS-HTK commented on .github/workflows/github_test_action.yml:
```diff
@@ -342,7 +345,7 @@ jobs:
         # Careful when copying this command. The PYTHONPATH setup is Linux specific syntax.
         run: |
-          pytest --nbmake --nbmake-timeout=900 --timeout=900 "./tutorials"
+          pytest --splits 3 --group ${{ matrix.group }} --nbmake --nbmake-timeout=900 --timeout=900 "./tutorials"
```
I know it has been forgotten on develop, but the matrix should also be implemented on the tutorial_warnings_tests, as they run a long time as well.

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
fixes from review

KS-HTK committed:
Merge branch 'release/v4.0.0' into structure_dict_extension

KS-HTK commented on pandapower/control/controller/trafo_control.py:
```diff
@@ -179,16 +183,14 @@ def set_recycle(self, net):
     def __repr__(self):
-        s = '%s of %s %s' % (self.__class__.__name__, self.element, self.element_index)
-        return s
+        return '%s of %s %s' % (self.__class__.__name__, self.element, self.element_index)
```
why not use f-String notation?

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -31,31 +33,31 @@ def create_gen(
+    vm_pu: float = get_default_value("gen", "vm_pu"),
     sn_mva: float = nan,
-    name: str | None = None,
+    name: str = pd.NA,
```
this should be `str | pd.NA` as pd.NA is not a str.

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -31,31 +33,31 @@ def create_gen(
-    curve_style: str | None = None,
-    controllable: bool | None = None,
+    scaling: float = get_default_value("gen", "scaling"),
+    type: str = pd.NA,
```
also should be `str | pd.NA`

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -31,31 +33,31 @@ def create_gen(
+    slack: bool = get_default_value("gen", "slack"),
+    id_q_capability_characteristic: int | None = pd.NA,
+    reactive_capability_curve: bool | None = None,
+    curve_style: str | None = pd.NA,
```
pd.NA is not None, is None allowed here? (same for line 49 and 52)

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -31,31 +33,31 @@ def create_gen(
-    power_station_trafo: int | float = nan,
-    in_service: bool = True,
-    slack_weight: float = 0.0,
+    power_station_trafo: int = pd.NA,
```
pd.NA is not int

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -187,21 +196,21 @@ def create_gens(
-    curve_style: str | Iterable[str] | None = None,
-    controllable: bool | Iterable[bool] | None = None,
+    scaling: float | Iterable[float] = get_default_value("gen", "scaling"),
+    type: str | Iterable[str] = pd.NA,
```
pd.NA is not str | Iterable[str]

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -187,21 +196,21 @@ def create_gens(
+    scaling: float | Iterable[float] = get_default_value("gen", "scaling"),
+    type: str | Iterable[str] = pd.NA,
+    slack: bool | Iterable[bool] = get_default_value("gen", "slack"),
+    id_q_capability_characteristic: Int | Iterable[Int] | None = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -187,21 +196,21 @@ def create_gens(
+    slack: bool | Iterable[bool] = get_default_value("gen", "slack"),
+    id_q_capability_characteristic: Int | Iterable[Int] | None = pd.NA,
+    reactive_capability_curve: bool | Iterable[bool] | None = None,
+    curve_style: str | Iterable[str] | None = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/gen_create.py:
```diff
@@ -187,21 +196,21 @@ def create_gens(
-    power_station_trafo: int | float = nan,
-    in_service: bool = True,
-    slack_weight: float = 0.0,
+    power_station_trafo: int = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/network_create.py:
```diff
@@ -16,7 +16,7 @@
 
 def create_empty_network(
-    name: str = "", f_hz: float = 50.0, sn_mva: float = 1, add_stdtypes: bool = True
+    name: str = "", f_hz: float = 50.0, sn_mva: float = 1, add_stdtypes: bool = True, structure: dict = None
```
None missing in typing.

If None should not be set by user use
```python
@overload
def create_empty_network(name: str = "", f_hz: float = 50.0, sn_mva: float = 1, add_stdtypes: bool = True) -> pandapowerNet: ...


@overload
def create_empty_network(name: str = "", f_hz: float = 50.0, sn_mva: float = 1, add_stdtypes: bool = True, structure: dict) -> pandapowerNet: ...

def create_empty_network(<…>
```

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -199,16 +196,16 @@ def create_transformers(
+    df: float | Iterable[float] = get_default_value("trafo", "df"),
     tap_changer_type: TapChangerWithTabularType | Iterable[str] | None = None,
-    tap_dependency_table: bool | Iterable[bool] = False,
+    tap_dependency_table: bool | Iterable[bool] = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -199,16 +196,16 @@ def create_transformers(
     id_characteristic_table: int | Iterable[int] | None = None,
     pt_percent: float | Iterable[float] = nan,
-    oltc: bool | Iterable[bool] = False,
+    # oltc: bool | Iterable[bool] = False,
```
remove? They will be in the kwargs if someone uses it. Check if this needs deprecation or is fine in the kwargs

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -256,18 +253,38 @@ def create_transformers(
+        tap_dependency_table=tap_dependency_table,
+        id_characteristic_table=id_characteristic_table,
+        pt_percent=pt_percent,
+        # oltc=oltc,
```
see above

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -292,21 +309,21 @@ def create_transformer_from_parameters(
-    oltc: bool = False,
-    tap_dependency_table: bool = False,
+    # oltc: bool = False,
+    tap_dependency_table: bool = pd.NA,
```
pd.NA missing in typing

and above.

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -524,21 +535,21 @@ def create_transformers_from_parameters(  # index missing ?
-    oltc: bool | Iterable[bool] = False,
-    tap_dependency_table: bool | Iterable[bool] = False,
+    # oltc: bool | Iterable[bool] = False,
+    tap_dependency_table: bool | Iterable[bool] = pd.NA,
```
pd.NA missing in typing
and above...

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -715,14 +724,14 @@ def create_transformer3w(
-    tap_at_star_point: bool = False,
-    tap_dependency_table: bool = False,
+    tap_at_star_point: bool = get_default_value("trafo3w", "tap_at_star_point"),
+    tap_dependency_table: bool = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -859,15 +865,15 @@ def create_transformers3w(
-    id_characteristic_table: int | Iterable[int] | None = None,
+    tap_at_star_point: bool | Iterable[bool] = get_default_value("trafo3w", "tap_at_star_point"),
+    tap_changer_type: float | Iterable[float] = nan,
+    tap_dependency_table: bool | Iterable[bool] = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -859,15 +865,15 @@ def create_transformers3w(
+    tap_at_star_point: bool | Iterable[bool] = get_default_value("trafo3w", "tap_at_star_point"),
+    tap_changer_type: float | Iterable[float] = nan,
+    tap_dependency_table: bool | Iterable[bool] = pd.NA,
+    id_characteristic_table: int | Iterable[int] = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -958,19 +993,19 @@ def create_transformer3w_from_parameters(
     vkr0_lv_percent: float = nan,
     vector_group: str | None = None,
-    tap_dependency_table: bool = False,
+    tap_dependency_table: bool = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -1142,30 +1175,30 @@ def create_transformers3w_from_parameters(  # no index ?
-    tap_side: HVMVLVType | Iterable[str] | None = None,
+    shift_mv_degree: float | Iterable[float] = get_default_value("trafo3w", "shift_mv_degree"),
+    shift_lv_degree: float | Iterable[float] = get_default_value("trafo3w", "shift_lv_degree"),
+    tap_side: HVMVLVType | Iterable[str] = pd.NA,
```
Does HVMVLVType include pd.NA?

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -1142,30 +1175,30 @@ def create_transformers3w_from_parameters(  # no index ?
-    tap_at_star_point: bool | Iterable[bool] = False,
-    tap_changer_type: float | Iterable[float] | None = None,
+    tap_at_star_point: bool | Iterable[bool] = pd.NA,
+    tap_changer_type: str | Iterable[str] = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/create/trafo_create.py:
```diff
@@ -1142,30 +1175,30 @@ def create_transformers3w_from_parameters(  # no index ?
-    id_characteristic_table: int | Iterable[int] | None = None,
+    vector_group: str | Iterable[str] = pd.NA,
+    tap_dependency_table: bool | Iterable[bool] = pd.NA,
+    id_characteristic_table: int | Iterable[int] = pd.NA,
```
pd.NA missing in typing

KS-HTK commented on pandapower/test/api/test_create.py:
The FIXME's in this file should be adressed

KS-HTK commented on pandapower/test/grid_equivalents/test_get_equivalent.py:
```diff
@@ -218,18 +218,16 @@ def test_basic_usecases(eq_type):
-        assert np.allclose(net1.bus.min_vm_pu.values,
-                           np.array([0.9, 0.91, np.nan, np.nan, 0.93]), equal_nan=True)
+        # assert np.allclose(net1.bus.min_vm_pu.to_numpy(), np.array([0.9, 0.91, np.nan, np.nan, 0.93]), equal_nan=True)
+        assert np.allclose(net1.bus.min_vm_pu.to_numpy(), np.array([0.9, 0.91, 0., 0., 0.93]))
```
the commented out assert should be removed

KS-HTK commented on pandapower/test/network_schema/tools/test_pandera_bus_index_validation.py:
```diff
@@ -0,0 +1 @@
+# TODO:
```
?

KS-HTK commented on pandapower/auxiliary.py:
```diff
@@ -86,6 +88,7 @@
 T = TypeVar("T")
 
 
+# FIXME: Remove this!
```
Can this be removed? If so remove.

mrifraunhofer committed:
ltds_pce (#2944)

* added PowerElectronicsConnection to the LTDS converter, updated LTDS notebooks (PowerElectronicsConnection)

* updated LTDS notebooks for mapping Excel data

mrifraunhofer committed:
ltds_notebook (#2949)

* updated ltds notebooks

* updated LTDS notebooks for mapping Excel data

* updated LTDS notebooks for mapping Excel data

mrifraunhofer committed:
ltds_notebook (#2950)

* updated ltds notebooks

* updated LTDS notebooks for mapping Excel data

* updated LTDS notebooks for mapping Excel data

* UKPN Jupiter Notebook

* fix UKPN Jupiter Notebook

vogt31337 committed:
Feature check vkr percent (#2948)

* added check to check if vkr_percent values are reasonable.

* added changelog entry.

---------

Co-authored-by: KS_HTK <2981026+KS-HTK@users.noreply.github.com>

vogt31337 committed:
Feature additional diagnostic functions (#2952)

* added more diagnostic functions.

* added changelog entry.

* fixed a small bug, used the wrong value in the reactance field.

arisaka97 committed:
Correct loading_percent description in motor_par.csv (#2947)

Updated the description of loading_percent.

pawellytaev committed:
ucte2pp conversion fixes (#2823)

* consider empty R dataframe in ucte conversion

* convert impedances and trafos with G/B=NaN to G/B=0

* change fillna for trafos and impedances to the ucte_parser

* fillna df if 0

* fix nan values in impedances and trafos

* ucte2pp: averaging voltages and clean up parser

* changelog entries for ucte2pp fixes

* average voltage setpoints using topological search instead of ucte name

* empty space at EOF

* fix codacy issue

* remove create_switch from pp_import_functions

* fix create_vsc functions

* changelog :)

* change symmetrical tap changer to 90 degree tap step degree

* correct handling of symmetrical PSTs in ucte2pp

* comment averaging function based on PR review

* changes ucte converter

* ucte2pp: add flag for voltage harmonization, with default false

* fix test

* fix tests and restrict tolerances a bit

---------

Co-authored-by: mrichter <mario.richter@iee.fraunhofer.de>
Co-authored-by: V3 <mike.vogt@iee.fraunhofer.de>

MoritzHauff committed:
Fix 2733 (#2951)

* change patch sizes for plot (#2733)

* fix patch sizes for plot only once (#2733)

`get_collection_sizes()` takes arguments to adjust the relative size of
an element.

---------

Co-authored-by: V3 <mike.vogt@iee.fraunhofer.de>

jerelang committed:
Fix runopp(init="results") so the OPF solver preserves the model initial vector (#2902)

* Fix runopp(init=results) warm-start handling in pips OPF

* Update CHANGELOG for upcoming release fixes and features

---------

Co-authored-by: jereOG <jerelang@users.noreply.github.com>
Co-authored-by: V3 <mike.vogt@iee.fraunhofer.de>

jwiemer112 committed:
Feature/highlight plot (#2833)

* integrated highlighting feature into simple_plot

* integrated hover feature into simple_plot

* update plotting tutorial

* adjust color

heckstrahler committed:
fixes for review

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
typing for for network_create

furqan463 committed:
Fix unit g0_us (#2958)

attribute name g0_us_per_km is incorrectly named as g0_nf_per_km in add_zero_impedance_parameters function.

KS-HTK committed:
Feature/create tests refactoring (#2960)

* refactored create tests

* fixed issue in bus_create

* added test placeholders
added some tests

* added missing tests for wards

* added test for create_line

* added tests for create_measurement
updated schema
removed checks in create_measurement when covered by typing
removed forced element_type to bus, there is a raise UserWarning below for this case.

* added test_create_gen

* added tests for create_motor
updated schema to be consisten with create function and other tests

* removed check_existing column and added pd.NA to allowed check

* added trafo create tests

* added create switch test
removed incorrect measurement test

* added create storage test

* added tests for create impedance
fixed issue in create_impedances

* fixed issue in test_create_pwl_costs

* added test_create_poly_costs

* added tests for ext_grid_create

* fixed issue in test_gen_create.py

* added tests for create_group

* added tests for create load

* added tests for create source dc

* fixed issues in vsc_stacked schema
fixed issue in schema validation where a forain bus check would run for custom columns

* fixed schema validation issues in test_line_create.py

* added tests for shunt_create

* fixed call for _bus_index_validation

* added missing tests for sgen create functions

* fixed issues resulting from schema modification

* fixed measurement.side dtype

* cleaned up storage test to reduce code duplication

* cleaned up switch test to reduce code duplication

* fixed types in test_line_create.py

* sonar maintainability improvements.

* fixed all failing tests

* updated pandera version for py3.14 support

* fixed typing version of pandera

* removed float comparison

* removed float comparison

* removed more float comparisons

* removed more float comparisons

* fixed missing import

* fixed more float comparisons

* fixed more float comparisons

* fixed remaining sonar warnings

KS-HTK committed:
updated branding with versions from marketing (#2953)

furqan463 committed:
[IMPROVEMENT] fix `bool` parameters input in `read_excel` and add `add_basic_std_types` parameter (#2961)

* Add basic std types flag to from_excel

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Fix reading bool

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Fix backward compatibility

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Silence sonarcloud and codacy

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Remove ruff ignore

---------

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

KS-HTK committed:
added permissions block to all workflows (#2964)

KS-HTK committed:
updated the issue templates to use the issue types instead of labels and title (#2966)

minor changes in forms layouts

KS-HTK committed:
Feature/revised issue templates (#2967)

* updated the issue templates to use the issue types instead of labels and title
minor changes in forms layouts

* fix for config not parsing by github (hopefully)
added documentation template
added emoji to template title for more visual guidance

* uniform formatting for templates

* added missing blank line at end of file

jwiemer112 committed:
display logger message only if GetActiveScenario() isn't None (#2969)

Only display a logger message if GetActiveScenario is not None
Added else case with a info message if no study case / scenario is active.

furqan463 committed:
Update docstring (#2970)

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

mrifraunhofer committed:
ltds_notebook (#2972)

* updated ltds notebooks

* updated LTDS notebooks for mapping Excel data

* updated LTDS notebooks for mapping Excel data

* UKPN Jupiter Notebook

* fix UKPN Jupiter Notebook

* fix UKPN Jupiter Notebook

* fix UKPN Jupiter Notebook

furqan463 committed:
Update lightsim2grid version (#2976)

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

nis-nurr committed:
fix cmin for low voltage 10% tolerance (#2913)

* fix(cmin): fix cmin for low voltage 10% tolerance

furqan463 committed:
[Fix] Fix `reindex_buses` against creating duplicate indecies (#2963)

* Fix reindex_buses

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Add allow_duplicate_index argument

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Fix unused import

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Fix test

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

---------

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

KS-HTK committed:
feature/doc strings update (#2979)

* added/reformatted many docstrings

* added show_plot=False to test_plotting_toolbox

* refactored docstrings and added types

* fixed f-string in groups.py

* fixed incorrect type

* fixed incorrect type

* fixed missing import

* fixed import issue with specific lightsim2grid versions

* fixed mypy issues in contingency
refactored test in test_contingency

* fixed mypy issues

* updated var name

KS-HTK committed:
updated the contributing.rst (#2973)

* updated the contributing.rst

* removed raw directive

* changed contributing to markdown for better github compatibility
replaced contributing on readTheDocs by link to GitHub

* Added a section on changelog to the contributing.md
Added a changelog entry to the changelog

* Update CHANGELOG.rst

jwiemer112 committed:
[Feature]: Toggle Colormap for simple_plot (#2971)

* colormap-integration in simple_plot

* adjust colors
*add docstrings
*add second button to better visualize options for the user

* change ext_grid color

* some formatting

* only plot colormap if buses or lines are available (fix tutorial-test error)

* removed some spaces
removed some whitelines
added results to hover info

* add changelog entry

* adepted authors

* adapted zorder
make plotting bus-switches optional

* typo

* small adaptions

* small adjustments

---------

Co-authored-by: KS_HTK <2981026+KS-HTK@users.noreply.github.com>

mrifraunhofer committed:
ltds_notebook (#2980)

* updated ltds notebooks

* updated LTDS notebooks for mapping Excel data

* updated LTDS notebooks for mapping Excel data

* UKPN Jupiter Notebook

* fix UKPN Jupiter Notebook

* fix UKPN Jupiter Notebook

* fix UKPN Jupiter Notebook

* update LTDS Jupiter Notebook respecting new Excel format

---------

Co-authored-by: David Heck <david.heck@iee.fraunhofer.de>

furqan463 committed:
Fix diagnostic report (#2983)

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

marcopau committed:
Tutorials for UKPN (#2984)

* Fix tutorial on power flow

* Updated Jupyter notebooks for UKPN

* Fix on tutorial directory

heckstrahler committed:
fix missing zipload parameters

KS-HTK committed:
Feature/structure tests fixes (#2986)

* fixed cim converter not setting const_[zi]_[pq]_percent columns

* fixed regression from merge

* disable voltage_depend_loads for lightsim2grid tests
added warning for ls2g & voltage_depend_loads in doc

* revert once voltage_depend_loads has been fixed.
this commit only disables voltage_depend_loads.

---------
Co-authored-by: David Heck <david.heck@iee.fraunhofer.de>

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

furqan463 committed:
[Fix] `DiscreteTapControl` to work with negative `tap_step_percent` (#2990)

* abs tap_step_percent

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Add changelog, test case

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

* Remove code duplication

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

---------

Signed-off-by: furqan463 <ahmadfurqanc@gmail.com>

heckstrahler committed:
remove not needed todo

heckstrahler committed:
Merge branch 'develop' into structure_dict_extension

heckstrahler committed:
Merge branch 'release/v4.0.0' into structure_dict_extension

KS-HTK committed:
added tests for bus_index_validation.py (#2994)

fixed incorrect docstring type

heckstrahler committed:
fix typing

heckstrahler committed:
add changelog entry
